library/ssl_srv.c

```diff
@@ -3320,30 +3458,126 @@ static int ssl_parse_encrypted_pms( mbedtls_ssl_context *ssl,
+                                            peer_pmssize ) );
+        case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+            ssl->handshake->async_in_progress = 1;
+            return( MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS );
+        default:
+            MBEDTLS_SSL_DEBUG_RET( 1, "f_async_decrypt_start", ret );
+            return( ret );
+        }
+    }
+#endif /* MBEDTLS_SSL_ASYNC_PRIVATE */
+
+    if( ! mbedtls_pk_can_do( private_key, MBEDTLS_PK_RSA ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "got no RSA private key" ) );
+        return( MBEDTLS_ERR_SSL_PRIVATE_KEY_REQUIRED );
+    }
+
+    ret = mbedtls_pk_decrypt( private_key, p, len,
+                              peer_pms, peer_pmslen, peer_pmssize,
+                              ssl->conf->f_rng, ssl->conf->p_rng );
+    return( ret );
+}
+
+static int ssl_parse_encrypted_pms( mbedtls_ssl_context *ssl,
+                                    const unsigned char *p,
+                                    const unsigned char *end,
+                                    size_t pms_offset )
+{
+    int ret;
+    unsigned char *pms = ssl->handshake->premaster + pms_offset;
+    unsigned char ver[2];
+    unsigned char fake_pms[48], peer_pms[48];
+    unsigned char mask;
+    size_t i;
+    size_t peer_pmslen = 0;
+    unsigned int diff;
+
+    /* Initialize the first two bytes of peer_pms to something that
+     * will cause diff below to be nonzero if they are not overwritten.
+     * In principle this should be useless because
+     * ssl_decrypt_encrypted_pms() will write to peer_pms. However, in
+     * case of an operational error, ssl_decrypt_encrypted_pms() may
+     * leave peer_pms unmodified. This gives the correct behavior anyway
+     * because diff will be nonzero as soon as ret is nonzero and the
+     * behavior of the program only depends on whether diff is zero.
+     * However some bits of diff may depend on the
+     * otherwise-uninitialized part of peer_pms, and a memory analyzer
+     * such as Valgrind may reasonably complain about this. Hence this
+     * explicit initialization. */
+    peer_pms[0] = peer_pms[1] = ~0;
+
+    ret = ssl_decrypt_encrypted_pms( ssl, p, end,
+                                     peer_pms,
+                                     &peer_pmslen,
+                                     sizeof( peer_pms ) );
+
+#if defined(MBEDTLS_SSL_ASYNC_PRIVATE)
+    if ( ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS )
+        return( ret );
+#endif /* MBEDTLS_SSL_ASYNC_PRIVATE */
+
     mbedtls_ssl_write_version( ssl->handshake->max_major_ver,
-                       ssl->handshake->max_minor_ver,
-                       ssl->conf->transport, ver );
+                               ssl->handshake->max_minor_ver,
+                               ssl->conf->transport, ver );
+
+    /* Avoid data-dependent branches while checking for invalid
+     * padding, to protect against timing-based Bleichenbacher-type
+     * attacks. */
+    diff  = (unsigned int) ret;
```

Comment:
The value returned by `ssl_decrypt_encrypted_pms` can indicate four different conditions:

* Successful decryption.
* Successful use of the private key, but the decrypted value was not padded correctly.
* Operational error, e.g. unable to communicate with an external cryptoprocessor.
* The asynchronous callback must be called again later (tested above).

Before this PR, operational errors were rare — out of memory, or hardware accelerator failure. With this PR, `ssl_decrypt_encrypted_pms` can call a user-provided callback which would typically be prone to operational errors.

There is a conflict here between reporting operational errors as such, which is really useful for failure monitoring and for debugging, and the security need of not distinguishing successful operation from invalid padding (Bleichenbacher attack).

@mpg @yanesca @sbutcher-arm I'm considering changing the interface of the decrypt callback to report invalid padding as a success and to have an extra output which would indicate whether the padding was correct. Then `diff` would only be based on this additional output, and this function could return early on an operational error. This wouldn't solve the issue of operational errors from the pk module. What's your opinion on this?

Reply:
@yanesca It's not just convenience of debugging during program development: something like running out of memory or a communication error is also swallowed. And on the security front, the current code is fragile: it does compare the return value from the async decrypt callback (to `MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS`). And we're relying on the decrypt callback to not have timing that depends on whether the padding is valid and whether an error occurred.

Rambling on...

The alternative I envision is to change the callback's interface from

    ret = f_async_resume( ssl, output_buffer, &actual_output_length, output_buffer_size );
    if( ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS ) return( ret ); // try again later
    // if ret != 0 then randomize output_buffer, with timing that doesn't depend on whether ret == 0

to

    operational_ret = f_async_resume( ssl, output_buffer, &actual_output_length, output_buffer_size, &padding_is_correct );
    if( operational_ret != 0 ) return( operational_ret ); // try again later, or not at all because there was an operational error
    // if !padding_is_correct then randomize output_buffer, with timing that doesn't depend on whether padding_is_correct

The problem is, how would one implement such a callback? For example, on top of PKCS#11, you'd have to check the return code from the token and split that into `operational_ret` vs `padding_is_correct`. That's just as hard to do in a time-constant way, and there's more of a risk of getting it seriously wrong if the developer writing the callback doesn't know exactly what they're doing.

I'm afraid that timing dependencies on padding correctness are really hard to avoid unless you control the whole ecosystem.

Padding is evil.

Reply:
Oh I see now, the application needs to know about those operational errors to act accordingly.

I don't see how differentiating between an operational error and other return codes would give the attacker the timing oracle he needs for Bleichenbacher.

Actually it can in some extreme and non-intuitive cases, but I would consider that a serious flaw in the token. And it is less likely that the token designers/engineers don't now what they are doing than the developer writing the callback.